### ISSUE TITLE
In case you miss it "Provide a method to know when HTML it ready for Javascript interactions like click()"

### ISSUE THREAD

**Issue by darbid** [ukandrewc/Webview2.Winforms.DOM#10]
I have just made a suggestion over in WebView2 which I am sure you will read, but just in case you do not here is the link;

[Provide a method to know when HTML it ready for Javascript interactions like click()](Provide a method to know when HTML it ready for Javascript interactions like click())

I will copy my text below in case the link is lost;

This is pretty much a long shot and may not be appropriate for the WebView2, but I feel the need to write it here because if you got a dollar every time someone on StackOverflow asked the question "How do I know when a page with AJAX/Frames is ready" you would be rich. I realize and am very impressed to see the DOMContentLoaded this is really helpful, thank you.

Further I notice that Chrome DevTools Protocol - DOM.childNodeInserted already has something like I am suggesting.

I do plan to also suggest it over at Webview2.Winforms.DOM as well.

In many examples you see people just setting a timer in native code to wait after a page is fully loaded before they continue to automate the page. Many did this with the Webbrowser control. The problem with this is obvious. How long do you wait?

Then you have people setting up some javascript with an interval that polls for an element. As javascript is single threaded I would argue that polling for something would slow the page/browser down. I am not a web developer and my java script skills are basic but that is my guess.

One of the best ways to wait for an element to appear is with a MutationObserver, however one issue with this is that you need to know what you are waiting for. In my example below I have added an element ID, but this is not always the case. From what I can see you need to wait until the insertion "settles" which is why I add a debounce function as well. Effectively the 2 functions below combined wait for element injections to start (in my example waits for what you expect) and then once the element injections slow down to a user define period it assumes the DOM is complete.

I dont know what the API method code should be and exactly what the parameter should be either. (1) Parameter would of course be the element which you expect the mutation to happen on (2) the debounce time to wait (3) Optional an element that you are expecting. The only other thing which is not shown below but what I do use is the ability to cancel the observer.

Finally I would agree that CallDevToolsProtocolMethodAsync and the receiver would be an alternative. I also agree that Javascript injection, as I already do is an option. I also agree that DOMContentLoaded is also pretty good except the timing can be difficult for frames.

    `var debounceFunction;
        docUploadObserver = new MutationObserver(function (mutations) {
        mutations.forEach(function (mutation) {
            //
            for (var i = 0; i < mutation.addedNodes.length; i++) {
                //
                if (debounceFunction) {
                    debounceFunction();
                } else {
                    var item = mutation.addedNodes[i];
                    //
                    if (item.id) {
                        //console.log('openMainTab - ID: ' + item.id.toString());
                        if (item.id.includes('some check id')) {

                            debounceFunction = dbDebounce(function () {
                                //HTML has finished loading
                            }, waitTime);
                        }
                    }
                }
            }
        });
    });`

A debounce function is pretty standard I think but here is mine;

    `function dbDebounce(func, wait, immediate) {
    var timeout;
    return function () {
    var context = this,
        args = arguments;
    var later = function () {
        timeout = null;
        if (!immediate) {
            func.apply(context, args);
        }
    };
    var callNow = immediate && !timeout;
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
    if (callNow) {
        func.apply(context, args);
     }
    };
    }`

**Comment by ukandrewc** [comment]
I don't think you can make any assumptions that would allow this to be generic. You would need to know specific details about what you are waiting for.
We do something similar in our production application, using mutationObserver. Waiting for no changes for 1s before getting data from the page. However we have been tripped up by some pages that start a 1s timer and so our page never appears to be fully loaded.

**Comment by darbid** [comment]
> I don't think you can make any assumptions that would allow this to be generic

Agreed. And most likely because of this and because of the likelihood that the resulting event could cause confusion it is not something for a Microsoft API. However, in my use case the site removes and inserts elements quite often and events like DomContentLoaded or Window.load are fake friends.  Further these events produce confusion (from what I can see on Stackoverflow) even though they are well documented.

I would call my suggestion a cost effective (no JS polling) and intelligent (no simple wait) solution.

>We do something similar in our production application, using mutationObserver. Waiting for no changes for 1s before getting data from the page. However we have been tripped up by some pages that start a 1s timer and so our page never appears to be fully loaded.

Have a look at my code with the debounce. It solves your 1s delay issue even without having an expected element.

> that makes it easier to determine when the page should be fully loaded

I would say that in today's browser/web pages there is not a fool proof, slam dunk guarantee that a page is fully loaded. My code at least gives you a "settled" DOM and builds in a little wait time as buffer.

By the way I would say that the main/only assumption I make with this is that once elements start changing that the browser continues to make the changes in one go and does not do it in stages.  I agree this may not always be true but in my use case it seems to always be.

**Comment by ukandrewc** [comment]
I had forgotten that we do have a DOMDocumentChanged() Event, which is raised when the DOM element count changes.
It's only an experimental feature, so will probably evolve.

It's probably not necessary to debounce this call because that is already done by the MutationObserver. It won't fire until changes have stopped being made for  while.

I do want to compare MutationObserver with the DevTools event at some point.
